packages/blocks/src/_common/copilot/service/copilot-config.ts

```diff
@@ -0,0 +1,206 @@
+import type { AllServiceKind, GetMethod, ServiceImpl } from './service-base.js';
+
+export type VendorConfig = {
+  id: string;
+  vendorKey: string;
+  name: string;
+  data: unknown;
+};
+type ServiceType = string;
+export type ServiceConfigMap = Record<
+  ServiceType,
+  {
+    vendorId: string;
+    implName: string;
+  }
+>;
+export type CopilotConfigDataType = {
+  feature: Record<string, ServiceConfigMap>;
+  vendors: VendorConfig[];
+};
+
+type VendorPack<T extends AllServiceKind> = {
+  vendor: VendorConfig;
+  impl: ServiceImpl<GetMethod<T>, unknown>;
+};
+
+/**
+ * A simple PRNG, using a linear congruential generator.
+ * reference link: https://en.wikipedia.org/wiki/Linear_congruential_generator
+ */
+class SimplePRNG {
+  constructor(private seed: number) {}
+
+  next() {
+    const a = 1664525;
+    const c = 1013904223;
+    const m = 2 ** 32;
+    this.seed = (a * this.seed + c) % m;
+    return this.seed / m;
+  }
+}
+
+function shuffleArray(array: number[], seed: number) {
+  const prng = new SimplePRNG(seed);
+  for (let i = array.length - 1; i > 0; i--) {
+    const j = Math.floor(prng.next() * (i + 1));
+    [array[i], array[j]] = [array[j], array[i]];
+  }
+}
+
+function decode(encodedStr: string, seed: number) {
+  const indexArray = Array.from(encodedStr).map((_, index) => index);
+  shuffleArray(indexArray, seed);
+  const inverseIndexArray = indexArray
+    .map((originalIndex, newIndex) => [originalIndex, newIndex])
+    .sort((a, b) => a[0] - b[0])
+    .map(pair => pair[1]);
+  return inverseIndexArray.map(index => encodedStr[index]).join('');
+}
+
+const seed = 1234567890;
+
+export class CopilotConfig {
+  _config?: CopilotConfigDataType;
+
+  loadFromLocalStorage() {
+    try {
+      return JSON.parse(localStorage.getItem('copilotConfig') ?? '');
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    } catch (e) {
+      return {};
+    }
+  }
+
+  loadVendorFromUrl() {
+    const result: VendorConfig[] = [];
+    const params = new URLSearchParams(window.location.search);
+    const openAIKey = params.get('openai');
+
+    if (openAIKey) {
+      result.push({
+        id: 'url-openai',
+        vendorKey: 'OpenAI',
+        name: 'url',
+        data: {
+          apiKey: 'sk-' + decode(openAIKey, seed),
```

Comment:
What's the point?

Reply:
Temporary solution to avoid the api-key appearing directly on the url, will be removed when this version is released.